numpy/core/src/multiarray/convert_datatype.c

```diff
@@ -968,12 +968,17 @@ PyArray_FindConcatenationDescriptor(
 }
 
 
-/*NUMPY_API
- * Produces the smallest size and lowest kind type to which both
- * input types can be cast.
+/*
+ * Same as `PyArray_PromoteTypes`, but allows falling back to an object result
+ * when no common DType is found.  This is currently only needed for the
+ * string and number promotion deprecation!
+ *
+ * TODO: After this deprecation is over, the `object_fallback` may well be
```

Comment:
Would be best if this could just change from "may well be useless!" to "should be removed"